compiler/rustc_hir_analysis/src/collect/predicates_of.rs

```diff
@@ -341,6 +320,51 @@ fn gather_explicit_predicates_of(tcx: TyCtxt<'_>, def_id: LocalDefId) -> ty::Gen
     }
 }
 
+fn bidirectional_lifetime_predicates<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    item_def_id: LocalDefId,
+    bidir_outlives: impl Iterator<Item = (Lifetime, (LocalDefId, Symbol, Span))>,
+) -> Vec<(Predicate<'tcx>, Span)> {
+    let mut predicates = Vec::new();
+    let icx = ItemCtxt::new(tcx, item_def_id);
+
+    let generics = tcx.generics_of(item_def_id.to_def_id());
```

Comment:
If I remove this call and pass `&Generics` to this function, given that `generics_of(item_def_id.to_def_id())` was already calculated on `gather_explicit_predicates_of`, the amount of instructions counts is going to be reduced by 2,000,000 which is more or less the difference we get in the perf results. Anyway, I think this is irrelevant and something that we shouldn't trade the better API of the function for this "perf" wins.

Reply:
If generics_of is already in the parent, then just pass it in.